### ISSUE TITLE
Local storage handling in Nodes

### ISSUE THREAD

**Issue by vishh** [kubernetes/kubernetes#30799]
As of today Kubelet can consume storage via multiple devices & partitions. 
[This proposal](https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/disk-accounting.md) on disk accounting lists some of the well understood storage configurations,
1. Root partition
2. Docker Images partition - primarily meant for device mapper storage driver

However there are various other means by which Kubelet can consume local storage. 
1. Volumes can exist in non-root partition `--root-dir`
2. Hostpath volumes can use any partition or device
3. `/var/log` could exist in non-root partition.
4. `/var/lib/docker` (docker root dir) can exist in a different partition.
There could be more options.

Why does kubelet care about storage configuration?
1. Kubelet provides storage accounting for volumes, pods and nodes.
2. Kubelet attempts to ensure that local storage doesn't get full and evicts pods if necessary to free up storage.
3. Kubelet in the future will support storage isolation and allow pods to express disk requests & limits.

To avoid complicating the kubelet, I have the following proposal.

Restrict Kubelet's non-volume storage usage to a single partition excepting image filesystem. Kubelet on bootup will verify that the above mentioned storage modalities are all landing up on the same partition. 
A typical use case for a user to customize storage is to provide additional storage for volumes or to work around a small root partition. 
If a user does not intend to use the root filesystem, then they are expected to make `/var/lib/docker`, `/var/log` and `--root-dir` to point to the same storage partition. 
Images can continue to be optionally in a separate partition. 

Alternate approaches include kubelet tracking individual devices and supporting accounting, evictions and isolation on all of them. This is very tedious.

Use cases of HostPath volumes include mapping in SSDs and other HDDs. Volumes interface can be expanded to provide access to local storage. Assuming that users will want to share storage devices, kubelet will have to track additional storage devices via volume plugin interface and let pods consume these additional storage devices via volumes.

Given that HostPath volumes can be shared, kubelet does not account for them. By expanding scope of volumes interface to all local storage devices, it will allow for accounting and isolation and avoid usage of HostPath volumes except for read-only use cases.
In the future, kubelet will have to expose these additional devices to the scheduler in-order to find the node that can fit a pod with `disk` requirements for containers and volumes.

There is still the problem of freeing up storage on additional storage devices / partitions exposed via volumes. One easy way to work around that would be to place all pods that do not have a disk limit on the primary partition that kubelet tracks. Volumes with limits do not require disk reclamation by kubelet.

By forcing a single partition, `disk` resources expressed on a container or pod can be easily handled by kubelet. `disk` resource on a volume can be handled by the volume directly. K8s metrics API need not have to deal with storage configuration of individual nodes. 

Consider this proposal as a draft. The main purpose is to start a discussion & come up with some high level design conclusions. 

cc @derekwaynecarr @thockin @smarterclayton @matchstick @saad-ali @pmorie

**Comment by vishh** [comment]
xref: #30782

**Comment by derekwaynecarr** [comment]
I think we need to potentially support tracking `/var/lib/docker` depending on storage driver.  For example, it seems highly encouraged best practice for `overlay` to use a separate device with own filesystem per https://docs.docker.com/engine/userguide/storagedriver/overlayfs-driver/#overlayfs-and-docker-performance

>>Unfortunately you can only specify the number of inodes in a filesystem at the time of creation. For this reason, you may wish to consider putting /var/lib/docker on a separate device with its own filesystem, or manually specifying the number of inodes when creating the filesystem.

**Comment by aronchick** [comment]
Terrific stuff! Can you please add this as a feature in the feature repo (1.5?)

**Comment by thockin** [comment]
I have expressed this before, and I'd be happy to be wrong, but I don't
think this will fly.

There are too many ways that storage can be set up to make this demand.
Even in GCE it does not seem unreasonable to add another PD to expand
storage capacity.  Now add multiple types of media.  Clouds support local
SSD which is not dynamically created/destroyed like EBS or PD.  Next will
be NVM or PCM or who knows what storage devices.  We will need a way to
express the desire to use these.

When I spoke to a couple customers about their expectations wrt SSD, I got
the opposite answer than I expected.  Assume you have an SSD.  It is very
fast and very expensive.  You need the speed.  Do you expect to get a
dedicated SSD device (fast, predicatble perf but expensive) or a shared SSD
(cheaper but unpredictable perf)?

So I would love to hear everyone say "this is good enough", but i just
don't buy it.

**Comment by vishh** [comment]
@thockin its possible that my description was unclear. Let me try again. The proposal is as follows:
* Kubelet will support best effort evictions only on one partition. All best effort usage of disk either via volumes or rootfs or logs will have to be on that parition
* All additional storage partitions will be dedicated for non best-effort scenarios, where the pod specifies a disk limit. Kubelet will continue to track usage and enforce limits on those secondary partitions, but it will not try to overcommit them. 

There is still the problem of disk IO isolation and guarantees which gets very tricky depending on the way these partitions get shared.

**Comment by vishh** [comment]
I talked to @thockin offline about this issue. I will come up with a short
proposal on local storage handling in k8s and post it to sig-node and
sig-storage in the next few weeks!

**Comment by derekwaynecarr** [comment]
@danmcp -- per our discussion today on supporting multiple storage drivers in a cluster where posix compliance is important.

**Comment by dims** [comment]
@vishh @saad-ali @calebamiles - switching this to 1.6 as it's too late for 1.5. ok? (please switch it right back if you disagree)

**Comment by krmayankk** [comment]
@vishh is there a proposal for local storage handling that you already shared somewhere as mentioned above ?

**Comment by jheiss** [comment]
@krmayankk I think that proposal may be the (just posted) kubernetes/community#306

**Comment by msau42** [comment]
@vishh, can we close this?  I think the current local storage design using PVs that we are implementing differs from this original idea.

**Comment by vishh** [comment]
@msau42 yes, this can be closed with appropriate pointers to proposals and PRs that resolve the issues mentioned here.

**Comment by msau42** [comment]
Proposal is kubernetes/community#306
PR tracker is #43640
